vendor/github.com/mendersoftware/mender-artifact/areader/reader.go

```diff
@@ -397,10 +385,19 @@ func (ar *Reader) readHeaderV3(version []byte) error {
 			return errors.Wrap(err, "readHeaderV3")
 		}
 		if validPath {
+			shouldBeSigned := false
+			if ar.ShouldBeSignedCallback != nil {
+				shouldBeSigned = ar.ShouldBeSignedCallback(ar.hInfo)
+			}
 			// Artifact should be signed, but isn't, so do not process the update.
-			if ar.shouldBeSigned && !ar.IsSigned {
+			if shouldBeSigned && !ar.IsSigned {
 				return errors.New("reader: expecting signed artifact, but no signature file found")
 			}
+			// First read and verify signature
+			if err = signatureReadAndVerify(ar.signature, ar.manifest.GetRaw(),
+				ar.VerifySignatureCallback, shouldBeSigned); err != nil {
```

Comment:
If the signature verification is done here, then if I read it correctly, it will be done multiple times, once for each file in the tar (likely only header and payload at the moment, but could increase if we implement multiple payloads). Do you actually need to move it out of its original location?

The same goes for the V2 variant.

Reply:
You're right, I originally did this because we didn't now if the payload needed to be signed (and with what key) until after reading the header. Since we removed the key scoping, we can roll this back to the original.